edb/server/pgsql/metaschema.py

```diff
@@ -919,6 +978,221 @@ def __init__(self):
+    """Get a JSON element by smallint index or raise an exception."""
+    text = r'''
+        SELECT
+            CASE jsonb_typeof(val)
+            WHEN 'object' THEN (
+                edgedb._raise_specific_exception(
+                    'wrong_object_type',
+                    'cannot index json ' || jsonb_typeof(val) ||
+                    ' by ' || pg_typeof(index)::text,
+                    det,
+                    NULL::jsonb
+                )
+            )
+            WHEN 'array' THEN (
+                edgedb._raise_exception_on_null(
+                    val -> index,
+                    'invalid_parameter_value',
+                    'json index ' || index::text || ' is out of bounds',
+                    det
+                )
+            )
+            ELSE
+                edgedb._raise_specific_exception(
+                    'wrong_object_type',
+                    'cannot index json ' || jsonb_typeof(val),
+                    det,
+                    NULL::jsonb
+                )
+            END
+    '''
+
+    def __init__(self):
+        super().__init__(
+            name=('edgedb', '_json_index'),
+            args=[('val', ('jsonb',)), ('index', ('smallint',)),
+                  ('det', ('text',))],
+            returns=('jsonb',),
+            volatility='immutable',
+            strict=True,
+            text=self.text)
+
+
+class JSONIndexByIntFunction(dbops.Function):
+    """Get a JSON element by int index or raise an exception."""
+    text = r'''
+        SELECT
+            CASE jsonb_typeof(val)
+            WHEN 'object' THEN (
+                edgedb._raise_specific_exception(
+                    'wrong_object_type',
+                    'cannot index json ' || jsonb_typeof(val) ||
+                    ' by ' || pg_typeof(index)::text,
+                    det,
+                    NULL::jsonb
+                )
+            )
+            WHEN 'array' THEN (
+                edgedb._raise_exception_on_null(
+                    val -> index,
+                    'invalid_parameter_value',
+                    'json index ' || index::text || ' is out of bounds',
+                    det
+                )
+            )
+            ELSE
+                edgedb._raise_specific_exception(
+                    'wrong_object_type',
+                    'cannot index json ' || jsonb_typeof(val),
+                    det,
+                    NULL::jsonb
+                )
+            END
+    '''
+
+    def __init__(self):
+        super().__init__(
+            name=('edgedb', '_json_index'),
+            args=[('val', ('jsonb',)), ('index', ('int',)),
+                  ('det', ('text',))],
+            returns=('jsonb',),
+            volatility='immutable',
+            strict=True,
+            text=self.text)
+
+
+class JSONIndexByBigintFunction(dbops.Function):
+    """Get a JSON element by bigint index or raise an exception."""
+    text = r'''
+        SELECT
```

Comment:
I'm thinking we shouldn't copy-paste the function body for all argument variants.  Please either make alternatives into small wrappers that call the main implementation with an up-casted argument, or cast at the call site.

Reply:
In this case it saves me from checking whether JSON indexing is done by some form of integer vs. text and upcasting the integer to `bigint` in Python. But I could move that logic from implicit Postgres polymorphism into Python entirely.